fs/glob_test.ts

```diff
@@ -0,0 +1,127 @@
+
+async function touch(path: string): Promise<void> {
+  await open(path, "w");
+}
+
+async function walkArray(
+  dirname: string = ".",
+  options: WalkOptions = {}
+): Promise<Array<string>> {
+  const arr: string[] = [];
+  for await (const f of walk(dirname, { ...options })) {
+    arr.push(f.path.replace(/\\/g, "/"));
+  }
+  arr.sort();
+  const arr_sync = Array.from(walkSync(dirname, options), (f: FileInfo) =>
+    f.path.replace(/\\/g, "/")
+  ).sort();
+  assert.equal(arr, arr_sync);
+  return arr;
+}
+
+test({
+  name: "glob: glob to regex",
+  fn() {
+    assert.equal(glob("unicorn.*")[0] instanceof RegExp, true);
+    assert.equal(glob("unicorn.*")[0].test("poney.ts"), false);
+    assert.equal(glob("unicorn.*")[0].test("unicorn.py"), true);
+    assert.equal(glob("*.ts")[0].test("poney.ts"), true);
+    assert.equal(glob("*.ts")[0].test("unicorn.js"), false);
+    assert.equal(
+      glob(join("unicorn", "**", "cathedral.ts"))[0].test(
+        join("unicorn", "in", "the", "cathedral.ts")
+      ),
+      true
+    );
+    assert.equal(
+      glob(join("unicorn", "**", "cathedral.ts"))[0].test(
+        join("unicorn", "in", "the", "kitchen.ts")
+      ),
+      false
+    );
+    assert.equal(
+      glob(join("unicorn", "**", "bathroom.*"))[0].test(
+        join("unicorn", "sleeping", "in", "bathroom.py")
+      ),
+      true
+    );
+    assert.equal(
+      glob(join("unicorn", "!(sleeping)", "bathroom.ts"), {
+        extended: true
+      })[0].test(join("unicorn", "flying", "bathroom.ts")),
+      true
+    );
+    assert.equal(
+      glob(join("unicorn", "(!sleeping)", "bathroom.ts"), {
+        extended: true
+      })[0].test(join("unicorn", "sleeping", "bathroom.ts")),
+      false
+    );
+  }
+});
+
+testWalk(
+  async (d: string) => {
+    await mkdir(d + "/a");
+    await touch(d + "/a/x.ts");
+  },
+  async function globInWalk() {
+    const arr = await walkArray(".", { match: glob("*.ts") });
+    assert.equal(arr.length, 1);
+    assert.equal(arr[0], "./a/x.ts");
+  }
+);
+
+testWalk(
+  async (d: string) => {
+    await mkdir(d + "/a");
+    await mkdir(d + "/b");
+    await touch(d + "/a/x.ts");
+    await touch(d + "/b/z.ts");
+    await touch(d + "/b/z.js");
+  },
+  async function globInWalkWildcardFiles() {
+    const arr = await walkArray(".", { match: glob("*.ts") });
+    assert.equal(arr.length, 2);
+    assert.equal(arr[0], "./a/x.ts");
+    assert.equal(arr[1], "./b/z.ts");
+  }
+);
+
+testWalk(
+  async (d: string) => {
+    await mkdir(d + "/a");
+    await mkdir(d + "/a/yo");
+    await touch(d + "/a/yo/x.ts");
+  },
+  async function globInWalkFolderWildcard() {
+    const arr = await walkArray(".", {
+      match: glob(join( "a", "**", "*.ts"), { flags:'g', extended:true, globstar: true })
+    });
+    assert.equal(arr.length, 1);
+    assert.equal(arr[0], "./a/yo/x.ts");
+  }
+);
+
+testWalk(
+  async (d: string) => {
+    await touch(d + "x.ts");
```

Comment:
these should be touch(d + "/x.ts")

This, and other glob tests, may not work on windows because of this...

Reply:
you're right. Changed it but now the assertion is wrong. The walk returns only 1 item which is:
`./x.js` and normally it would return `./x.js` also.

Reply:
Something is definitely weird. I think it's in that include function, in the if statement, but was unable to nail it down / work out what is causing it.

Reply:
Is the pattern somehow mutable? I get these being different!!
```ts
  const a = options.match && (!options.match.some(pattern => pattern.test(path)));
  const b = options.match && (!options.match.some(pattern => pattern.test(path)));
  const c = options.match && (!options.match.some(pattern => pattern.test(path)));
  console.log(a, b, c)
```

Reply:
This is due to the `g` flag.
See: https://stackoverflow.com/questions/1520800/why-does-a-regexp-with-global-flag-give-wrong-results

So in this case we might have to change this part in globrex:
```typescript
    if (!flags.includes('g')) {
        regex = `^${regex}$`;
        segment = `^${segment}$`;
        if (filepath) path.regex = `^${path.regex}$`;
    }
```
To only get the generated regex pattern. Like returning the raw string expression and cast it by ourselves in the glob function.

Reply:
> This is like Hitchhiker's Guide to the Galaxy API design here. "That pitfall that you fell in has been perfectly documented in the spec for several years, if you had only bothered to check" 

very cool! 

Not sure whether that's best way to fix, or if should use the lastIndex reset after each use.

Reply:
https://github.com/denoland/deno_std/pull/219/files#diff-f7a9ec5a7cf0d6e9d8822f7ab61bf494R100

added it and supposed to work. Trying to figure out with the assertion is wrong.

for example this script:
```javascript
var arr = ["x.js", "x.ts", "z.js"];
console.log('array test without')
let arrReg = [RegExp(/^x\..*$/, "g")]
arr.forEach(path => {
  let r = arrReg.some(pattern => {
    var f = pattern.test(path);
    return f;
  });
  console.log(`${path}:${r}`);
});

console.log('array test with')
let arrReg2 = [RegExp(/^x\..*$/, "g")]
arr.forEach(path => {
  let r = arrReg2.some(pattern => {
    var f = pattern.test(path);
    pattern.lastIndex = 0;
    return f;
  });
  console.log(`${path}:${r}`);
});
```
outputs
```
array test without
x.js:true
x.ts:false
z.js:false
array test with
x.js:true
x.ts:true
z.js:false
```